flink-clients/src/test/java/org/apache/flink/client/python/PythonUtilTest.java

```diff
@@ -0,0 +1,159 @@
+		String targetTmpDir = System.getProperty("java.io.tmpdir") +
+			File.separator + "target_" + UUID.randomUUID();
+
+		sourceTmpDirPath = new Path(sourceTmpDir);
+		targetTmpDirPath = new Path(targetTmpDir);
+		try {
+			sourceFs = sourceTmpDirPath.getFileSystem();
+			if (sourceFs.exists(sourceTmpDirPath)) {
+				sourceFs.delete(sourceTmpDirPath, true);
+			}
+			sourceFs.mkdirs(sourceTmpDirPath);
+			targetFs = targetTmpDirPath.getFileSystem();
+			if (targetFs.exists(targetTmpDirPath)) {
+				targetFs.delete(targetTmpDirPath, true);
+			}
+			targetFs.mkdirs(targetTmpDirPath);
+		} catch (IOException e) {
+			throw new RuntimeException("initial PythonUtil test environment failed");
+		}
+	}
+
+	@Test
+	public void testCopyPyflinkLibToTarget() {
+		try {
+			Path aFilePath = new Path(sourceTmpDirPath, "a.txt");
+			sourceFs.create(aFilePath, FileSystem.WriteMode.OVERWRITE);
+			Path targetFilePath = new Path(targetTmpDirPath, "a.txt");
+			PythonUtil.copyPyflinkLibToTarget(aFilePath.toString(), targetFilePath.toString());
+			Assert.assertTrue(targetFs.exists(targetFilePath));
+			sourceFs.delete(aFilePath, true);
+			targetFs.delete(targetFilePath, true);
+		} catch (IOException e) {
+			throw new RuntimeException("test copy lib to tmp dir failed.");
+		}
+	}
+
+	@Test
+	public void testCopy() {
+		Path aDirPath = new Path(sourceTmpDirPath, "a");
+		Path aFilePath = new Path(aDirPath, "a.txt");
+		Path aSubDirPath = new Path(aDirPath, "sub");
+		Path aSubFilePath = new Path(aSubDirPath, "sub.txt");
+		try {
+			sourceFs.mkdirs(aSubDirPath);
+			sourceFs.create(aFilePath, FileSystem.WriteMode.OVERWRITE);
+			sourceFs.create(aSubFilePath, FileSystem.WriteMode.OVERWRITE);
+			Path targetDirPath = new Path(targetTmpDirPath, "a");
+			PythonUtil.copy(aDirPath, targetDirPath);
+			Path targetFilePath = new Path(targetDirPath, "a.txt");
+			Path targetSubDirPath = new Path(targetDirPath, "sub");
+			Path targetSubFilePath = new Path(targetSubDirPath, "sub.txt");
+			Assert.assertTrue(targetFs.exists(targetDirPath));
+			Assert.assertTrue(targetFs.exists(targetFilePath));
+			Assert.assertTrue(targetFs.exists(targetSubDirPath));
+			Assert.assertTrue(targetFs.exists(targetSubFilePath));
+			sourceFs.delete(aDirPath, true);
+			targetFs.delete(targetDirPath, true);
+		} catch (IOException e) {
+			throw new RuntimeException("test copy source path to tmp dir failed " + e.getMessage());
+		}
+	}
+
+	@Test
+	public void testStartPythonProcess() {
+		PythonUtil.PythonEnvironment pythonEnv = new PythonUtil.PythonEnvironment();
+		pythonEnv.workingDirectory = targetTmpDirPath.toString();
+		pythonEnv.pythonPath = targetTmpDirPath.toString();
+		List<String> commands = new ArrayList<>();
+		Path pyPath = new Path(targetTmpDirPath, "word_count.py");
+		try {
+			targetFs.create(pyPath, FileSystem.WriteMode.OVERWRITE);
+			File pyFile = new File(pyPath.toString());
+			String pyProgram = "#!/usr/bin/python\n" +
```

Comment:
Would it be better that testing with a real pyflink program?

Reply:
Yeah.If relevant code have a it case,it would be better.But I don't have a good idea how to do this.If you have a good design, you can share with me.Thank you.

Reply:
I agree that there is actually no good way to run the real pyflink program because the flink dist file does not exist when this test is running. For this test I think it is ok to remain as it is.